File: test/model/spreadsheet/XLSXCorrectorTest.java
package model.spreadsheet;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

import model.spreadsheet.XLSXCorrector;

import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.apache.poi.xssf.usermodel.XSSFCell;
import org.junit.Test;

public class XLSXCorrectorTest {
  
  private XLSXCorrector corrector = new XLSXCorrector();
  private Path standardDocument = Paths.get("test/resources/spreadsheet/standard.xlsx");
  
  private Path schullandheimDir = Paths.get("test/resources/spreadsheet/schullandheim");
  private Path schullandheimMuster = Paths.get(schullandheimDir.toString(), "Aufgabe_Schullandheim_Muster.xlsx");
  private Path schullandheimTeilLoesung = Paths.get(schullandheimDir.toString(), "Aufgabe_Schullandheim.xlsx");
  
  @Test
  public void testCompareCellValues() {
    Workbook muster = corrector.loadDocument(schullandheimMuster);
    Workbook teilLsg = corrector.loadDocument(schullandheimTeilLoesung);
    
    XSSFCell musterCell = (XSSFCell) muster.getSheetAt(2).getRow(15).getCell(7);
    XSSFCell compareCell = (XSSFCell) teilLsg.getSheetAt(2).getRow(15).getCell(7);
    assertThat(corrector.compareCellValues(musterCell, compareCell), equalTo("Wert richtig."));
    
    // Wert in Muster, Compare leer
    musterCell = (XSSFCell) muster.getSheetAt(3).getRow(16).getCell(5);
    compareCell = (XSSFCell) teilLsg.getSheetAt(3).getRow(16).getCell(5);
    assertThat(corrector.compareCellValues(musterCell, compareCell), equalTo("Keinen Wert angegeben!"));
    
    // Wert in Muster, Compare falsch
    musterCell = (XSSFCell) muster.getSheetAt(3).getRow(19).getCell(5);
    compareCell = (XSSFCell) teilLsg.getSheetAt(3).getRow(19).getCell(5);
    // TODO: Zahl schoener formatieren... 12.142857142857142
    assertThat(corrector.compareCellValues(musterCell, compareCell),
        equalTo("Wert falsch. Erwartet wurde '12.142857142857142'."));
  }
  
  @Test
  public void testCompareCellFormulas() {
    Workbook muster = corrector.loadDocument(schullandheimMuster);
    Workbook teilLsg = corrector.loadDocument(schullandheimTeilLoesung);
    
    XSSFCell musterCell = (XSSFCell) muster.getSheetAt(2).getRow(15).getCell(7);
    XSSFCell compareCell = (XSSFCell) teilLsg.getSheetAt(2).getRow(15).getCell(7);
    assertThat(corrector.compareCellFormulas(musterCell, compareCell), equalTo("Formel richtig."));
    
    // Kein Formel in Muster, Compare leer
    musterCell = (XSSFCell) muster.getSheetAt(3).getRow(16).getCell(1);
    compareCell = (XSSFCell) teilLsg.getSheetAt(3).getRow(16).getCell(1);
    assertThat(corrector.compareCellFormulas(musterCell, compareCell), equalTo("Es war keine Formel anzugeben."));
    
    // Wert in Muster, Compare leer
    musterCell = (XSSFCell) muster.getSheetAt(3).getRow(16).getCell(5);
    compareCell = (XSSFCell) teilLsg.getSheetAt(3).getRow(16).getCell(5);
    assertThat(corrector.compareCellFormulas(musterCell, compareCell), equalTo("Keine Formel angegeben!"));
    
    // Wert in Muster, Compare leer
    musterCell = (XSSFCell) muster.getSheetAt(3).getRow(19).getCell(5);
    compareCell = (XSSFCell) teilLsg.getSheetAt(3).getRow(19).getCell(5);
    assertThat(corrector.compareCellFormulas(musterCell, compareCell), equalTo("Formel falsch. Der Bereich D20 fehlt."));
  }
  
  @Test
  public void testGetColoredRange() {
    Workbook muster = corrector.loadDocument(schullandheimMuster);
    List<XSSFCell> coloredRange = corrector.getColoredRange(muster.getSheetAt(1));
    assertTrue(coloredRange.isEmpty());
    
    coloredRange = corrector.getColoredRange(muster.getSheetAt(2));
    assertThat(coloredRange.size(), equalTo(28));
    for(int i = 0; i < 25; i++) {
      Cell cell = coloredRange.get(i);
      assertThat(cell.getRowIndex(), equalTo(15 + i));
      assertThat(cell.getColumnIndex(), equalTo(7));
    }
    Cell maennlich = coloredRange.get(26);
    assertThat(maennlich.getRowIndex(), equalTo(44));
    assertThat(maennlich.getColumnIndex(), equalTo(1));
    Cell weiblich = coloredRange.get(27);
    assertThat(weiblich.getRowIndex(), equalTo(45));
    assertThat(weiblich.getColumnIndex(), equalTo(1));
  }
  
  @Test
  public void testLoadDocument() {
    Workbook document = null;
    
    document = corrector.loadDocument(standardDocument);
    assertNotNull("Dokument standardDokument konnte nicht geladen werden!", document);
    
    document = corrector.loadDocument(schullandheimMuster);
    assertNotNull("Dokument schullandheimMuster konnte nicht geladen werden!", document);
    
    document = corrector.loadDocument(schullandheimTeilLoesung);
    assertNotNull("Dokument schullandheimMuster konnte nicht geladen werden!", document);
  }
  
  @Test
  public void testLoadDocumentWithWrongPath() {
    assertNull(corrector.loadDocument(Paths.get("")));
  }
  
  @Test
  public void testGetSheetCount() {
    Workbook document = null;
    
    document = corrector.loadDocument(standardDocument);
    assertNotNull("Standarddokument konnte nicht geladen werden!", document);
    assertThat(corrector.getSheetCount(document), equalTo(1));
    
    document = corrector.loadDocument(schullandheimMuster);
    assertThat(corrector.getSheetCount(document), equalTo(8));
    
    document = corrector.loadDocument(schullandheimTeilLoesung);
    assertThat(corrector.getSheetCount(document), equalTo(8));
  }
  
  @Test
  public void testCompareNumberOfChartsInDocument() {
    Workbook document = corrector.loadDocument(standardDocument);
    assertNull(corrector.compareNumberOfChartsInDocument(document, document));
  }
  
  @Test
  public void testGetSheetByIndex() {
    Workbook document = corrector.loadDocument(standardDocument);
    assertNotNull("Standarddokument konnte nicht geladen werden!", document);
    Sheet sheet = corrector.getSheetByIndex(document, 0);
    assertNotNull(sheet);
    
    Workbook muster = corrector.loadDocument(schullandheimMuster);
    Sheet musterSheet = corrector.getSheetByIndex(muster, 0);
    assertNotNull(musterSheet);
    assertThat(musterSheet.getRow(0).getCell(0).toString(), equalTo("Verwalten und Auswerten von Daten"));
    musterSheet = corrector.getSheetByIndex(muster, 4);
    assertNotNull(musterSheet);
    assertThat(musterSheet.getRow(15).getCell(10).toString(), equalTo("5% Rabatt auf den Gesamtpreis"));
  }
  
  @Test
  public void testCompareSheet() {
    // FIXME: implement and test!
    // fail("Not yet implemented");
  }
  
  @Test
  public void testSaveCorrectedSpreadsheet() {
    // FIXME: implement and test!
    // fail("Not yet implemented");
  }
  
  @Test
  public void testCloseDocument() {
    Workbook document = corrector.loadDocument(schullandheimMuster);
    assertNotNull(document);
    corrector.closeDocument(document);
  }
  
  @Test
  public void testGetCellByPosition() {
    Workbook muster = corrector.loadDocument(schullandheimMuster);
    Sheet musterSheet = corrector.getSheetByIndex(muster, 0);
    assertThat(corrector.getCellByPosition(musterSheet, 0, 0).toString(), equalTo(musterSheet.getRow(0).getCell(0)
        .toString()));
    
    assertNull(corrector.getCellByPosition(musterSheet, 7, 5));
    
    Sheet musterSheet4 = corrector.getSheetByIndex(muster, 4);
    assertThat(corrector.getCellByPosition(musterSheet4, 15, 10).toString(), equalTo(musterSheet4.getRow(15)
        .getCell(10).toString()));
    
  }
  
  @Test
  public void testSetCellComment() {
    String message = "Dies ist eine Testnachricht!";
    String message2 = "Dies ist eine zweite  Testnachricht!";
    
    Workbook document = corrector.loadDocument(standardDocument);
    XSSFCell cell = (XSSFCell) document.getSheetAt(0).getRow(0).getCell(0);
    
    assertNull(cell.getCellComment());
    
    corrector.setCellComment(cell, null);
    assertNull(cell.getCellComment());
    
    corrector.setCellComment(cell, "");
    assertNull(cell.getCellComment());
    
    corrector.setCellComment(cell, message);
    assertThat(cell.getCellComment().getString().getString(), equalTo(message));
    
    corrector.setCellComment(cell, message2);
    assertThat(cell.getCellComment().getString().getString(), equalTo(message2));
    
    cell.removeCellComment();
  }
  
  @Test
  public void testCompareChartsInSheet() {
    
    Workbook muster = corrector.loadDocument(schullandheimMuster);
    assertThat(corrector.compareChartsInSheet(muster.getSheetAt(0), muster.getSheetAt(0)),
        equalTo("Es waren keine Diagramme zu erstellen."));
    
    Workbook solution = corrector.loadDocument(schullandheimTeilLoesung);
    assertThat(corrector.compareChartsInSheet(solution.getSheetAt(0), muster.getSheetAt(2)),
        equalTo("Falsche Anzahl an Diagrammen im Sheet (Erwartet: 2, Gefunden: 0)."));
    
    assertThat(corrector.compareChartsInSheet(muster.getSheetAt(2), muster.getSheetAt(2)),
        equalTo("Diagramm(e) richtig."));
    // FIXME: Implement!
    // fail("Still things to implement!");
  }
}

File: modules/spread/test/model/XLSXCorrectorTest.scala
package model

import java.nio.file.{Files, Path, Paths}

import model.CommonUtils.RicherTry
import model.StringConsts._
import model.XLSXCorrectorTest._
import org.apache.poi.ss.usermodel.{Cell, Sheet, Workbook}
import org.apache.poi.xssf.usermodel.XSSFCell
import org.hamcrest.CoreMatchers.equalTo
import org.junit.Assert._
import org.junit._
import org.scalatest.Matchers._

import scala.util.{Failure, Success}

object XLSXCorrectorTest {

  println(System.getProperty("user.dir"))

  val testDir: Path = Paths.get("test", "resources", "spreadsheet")

  val stdDoc: Path = Paths.get(testDir.toString, "standard.xlsx")

  val testMuster: Path = Paths.get(testDir.toString, "schullandheim", "Aufgabe_Schullandheim_Muster.xlsx")
  val testSolution: Path = Paths.get(testDir.toString, "schullandheim", "Aufgabe_Schullandheim.xlsx")

  val testMusterCopy: Path = Paths.get(testDir.toString, "schullandheim", "testMusterCopy.xlsx")
  val testSolutionCopy: Path = Paths.get(testDir.toString, "schullandheim", "testSolutionCopy.xlsx")

  @BeforeClass
  def setUp(): Unit = {
    Files.copy(testMuster, testMusterCopy)
    Files.copy(testSolution, testSolutionCopy)
  }

  @AfterClass
  def tearDown(): Unit = {
    Files.delete(testMusterCopy)
    Files.delete(testSolutionCopy)
  }

}

class XLSXCorrectorTest {

  private def failureLoad(path: Path, e: Throwable): String = s"Could not load workbook $path:"

  @Test
  def testCloseDocument(): Unit = XLSXCorrector.loadDocument(testMusterCopy) match {
    case Failure(e) => Assert.fail(failureLoad(testMusterCopy, e))
    case Success(document) => XLSXCorrector.closeDocument(document)
  }

  private def getCell(workbook: Workbook, sheetIndex: Int, rowIndex: Int, cellIndex: Int) =
    workbook.getSheetAt(sheetIndex).getRow(rowIndex).getCell(cellIndex).asInstanceOf[XSSFCell]

  @Test
  def testCompareCellFormulas(): Unit = XLSXCorrector.loadDocument(testMusterCopy).zip(XLSXCorrector.loadDocument(testSolution)) match {
    case Failure(e) => Assert.fail(failureLoad(null, e))
    case Success((muster, teilLsg)) =>
      XLSXCorrector.compareCellFormulas(getCell(muster, 2, 15, 7), getCell(teilLsg, 2, 15, 7)) shouldBe(true, COMMENT_FORMULA_CORRECT)

      // Kein Formel in Muster, Compare leer
      XLSXCorrector.compareCellFormulas(getCell(muster, 3, 16, 1), getCell(teilLsg, 3, 16, 1)) shouldBe(false, COMMENT_FORMULA_FALSE)

      // Wert in Muster, Compare leer
      XLSXCorrector.compareCellFormulas(getCell(muster, 3, 16, 5), getCell(teilLsg, 3, 16, 5)) shouldBe(false, COMMENT_FORMULA_MISSING)

      // Wert in Muster, Compare leer
      XLSXCorrector.compareCellFormulas(getCell(muster, 3, 19, 5), getCell(teilLsg, 3, 19, 5)) shouldBe(false, "Formel falsch. Die Bereiche [D20] fehlen.")
  }

  @Test
  def testCompareCellValues(): Unit = XLSXCorrector.loadDocument(testMusterCopy).zip(XLSXCorrector.loadDocument(testSolution)) match {
    case Failure(e) => Assert.fail(failureLoad(null, e))
    case Success((muster, teilLsg)) =>

      XLSXCorrector.compareCellValues(getCell(muster, 2, 15, 7), getCell(teilLsg, 2, 15, 7)) shouldBe(true, COMMENT_VALUE_CORRECT)

      // Wert in Muster, Compare leer
      XLSXCorrector.compareCellValues(getCell(muster, 3, 16, 5), getCell(teilLsg, 3, 16, 5)) shouldBe(false, COMMENT_VALUE_MISSING)

      // Wert in Muster, Compare falsch
      // TODO: Zahl schoener formatieren... 12.142857142857142
      XLSXCorrector.compareCellValues(getCell(muster, 3, 19, 5), getCell(teilLsg, 3, 19, 5)) shouldBe(false, "Wert falsch. Erwartet wurde '12.142857142857142'.")
  }

  @Test
  def testCompareChartsInSheet(): Unit = XLSXCorrector.loadDocument(testMusterCopy).zip(XLSXCorrector.loadDocument(testSolution)) match {
    case Failure(e) => Assert.fail(failureLoad(null, e))
    case Success((muster, teilLsg)) =>
      XLSXCorrector.compareChartsInSheet(muster.getSheetAt(0), muster.getSheetAt(0)) shouldBe(false, COMMENT_CHART_FALSE)

      XLSXCorrector.compareChartsInSheet(teilLsg.getSheetAt(0), muster.getSheetAt(2)) shouldBe(false, "Falsche Anzahl Diagramme im Dokument (Erwartet: 2, Gefunden: 0).")

      XLSXCorrector.compareChartsInSheet(muster.getSheetAt(2), muster.getSheetAt(2)) shouldBe(true, COMMENT_CHART_CORRECT)
    // TODO: Implement!
    // Assert.fail("Still things to implement!")
  }

  @Test
  def testCompareNumberOfChartsInDocument(): Unit = XLSXCorrector.loadDocument(stdDoc) match {
    case Failure(e) => Assert.fail(failureLoad(stdDoc, e))
    case Success(document) => XLSXCorrector.compareNumberOfChartsInDocument(document, document) shouldBe(true, "")
  }

  @Test
  def testCompareSheet(): Unit = {
    // TODO: implement and test!
    // Assert.fail("Not yet implemented")
  }

  @Test
  def testGetCellByPosition(): Unit = XLSXCorrector.loadDocument(testMusterCopy) match {
    case Failure(e) => Assert.fail(failureLoad(testMusterCopy, e))
    case Success(muster) =>
      val musterSheet: Sheet = XLSXCorrector.getSheetByIndex(muster, 0)
      assertThat(XLSXCorrector.getCellByPosition(musterSheet, 0, 0).toString, equalTo(musterSheet.getRow(0).getCell(0).toString))

      assertNull(XLSXCorrector.getCellByPosition(musterSheet, 7, 5))

      val musterSheet4: Sheet = XLSXCorrector.getSheetByIndex(muster, 4)
      assertThat(XLSXCorrector.getCellByPosition(musterSheet4, 15, 10).toString, equalTo(musterSheet4.getRow(15).getCell(10).toString))

  }

  @Test
  def testGetColoredRange(): Unit = XLSXCorrector.loadDocument(testMusterCopy) match {
    case Failure(e) => Assert.fail(failureLoad(testMusterCopy, e))
    case Success(muster) =>
      assertTrue(XLSXCorrector.getColoredRange(muster.getSheetAt(1)).isEmpty)

      val coloredRange = XLSXCorrector.getColoredRange(muster.getSheetAt(2))
      coloredRange.size shouldBe (28)
      for (i <- 0 until 25) {
        val cell: Cell = coloredRange(i)
        cell.getRowIndex shouldBe (15 + i)
        cell.getColumnIndex shouldBe (7)
      }

      val maennlich: Cell = coloredRange(26)
      maennlich.getRowIndex shouldBe (44)
      maennlich.getColumnIndex shouldBe (1)

      val weiblich: Cell = coloredRange(27)
      weiblich.getRowIndex shouldBe (45)
      weiblich.getColumnIndex shouldBe (1)
  }

  @Test
  def testGetSheetByIndex(): Unit = {
    XLSXCorrector.loadDocument(stdDoc) match {
      case Failure(e) => Assert.fail(failureLoad(stdDoc, e))
      case Success(document) =>
        assertNotNull("Standarddokument konnte nicht geladen werden!", document)
        assertNotNull(XLSXCorrector.getSheetByIndex(document, 0))
    }

    XLSXCorrector.loadDocument(testMusterCopy) match {
      case Failure(e) => Assert.fail(failureLoad(testMusterCopy, e))
      case Success(muster) =>
        val musterSheet1: Sheet = XLSXCorrector.getSheetByIndex(muster, 0)
        assertNotNull(musterSheet1)
        musterSheet1.getRow(0).getCell(0).toString shouldBe ("Verwalten und Auswerten von Daten")

        val musterSheet2 = XLSXCorrector.getSheetByIndex(muster, 4)
        assertNotNull(musterSheet2)
        musterSheet2.getRow(15).getCell(10).toString shouldBe ("5% Rabatt auf den Gesamtpreis")
    }
  }

  @Test
  def testGetSheetCount(): Unit = {
    XLSXCorrector.loadDocument(stdDoc) match {
      case Failure(e) => Assert.fail(failureLoad(stdDoc, e))
      case Success(document) =>
        assertNotNull("Standarddokument konnte nicht geladen werden!", document)
        XLSXCorrector.getSheetCount(document) shouldBe (1)
    }

    XLSXCorrector.loadDocument(testMusterCopy) match {
      case Failure(e) => Assert.fail(failureLoad(testMusterCopy, e))
      case Success(document) =>
        XLSXCorrector.getSheetCount(document) shouldBe (8)
    }

    XLSXCorrector.loadDocument(testSolutionCopy) match {
      case Failure(e) => Assert.fail(failureLoad(testSolutionCopy, e))
      case Success(document) =>
        XLSXCorrector.getSheetCount(document) shouldBe (8)
    }
  }

  @Test
  def testLoadDocument(): Unit = {
    assertTrue("Dokument standardDokument konnte nicht geladen werden!", XLSXCorrector.loadDocument(stdDoc).isSuccess)

    assertTrue("Dokument schullandheimMuster konnte nicht geladen werden!", XLSXCorrector.loadDocument(testMusterCopy).isSuccess)

    assertTrue("Dokument schullandheimMuster konnte nicht geladen werden!", XLSXCorrector.loadDocument(testSolutionCopy).isSuccess)
  }


  def testLoadDocumentWithWrongPath(): Unit = assertTrue(XLSXCorrector.loadDocument(Paths.get("")).isFailure)

  @Test
  def testSaveCorrectedSpreadsheet() {
    // TODO: implement and test!
    // Assert.fail("Not yet implemented")
  }

  @Test
  def testSetCellComment(): Unit = XLSXCorrector.loadDocument(stdDoc) match {
    case Failure(e) => Assert.fail(failureLoad(stdDoc, e))
    case Success(document) =>
      val (message, message2) = ("Dies ist eine Testnachricht!", "Dies ist eine zweite  Testnachricht!")

      val cell: XSSFCell = getCell(document, 0, 0, 0)

      assertNull(cell.getCellComment)

      XLSXCorrector.setCellComment(cell, null)
      assertNull(cell.getCellComment)

      XLSXCorrector.setCellComment(cell, "")
      assertNull(cell.getCellComment)

      XLSXCorrector.setCellComment(cell, message)
      cell.getCellComment.getString.getString shouldBe (message)

      XLSXCorrector.setCellComment(cell, message2)
      cell.getCellComment.getString.getString shouldBe (message2)

      cell.removeCellComment()
  }
}
